### DISCUSSION TITLE
Representation*Event structure

### DISCUSSION THREAD

**Discussion by JWMB** [simpleidserver/SimpleIdServer#246] [Ideas]
Hi,
first off, thanks for your impressive work on these libraries, they've been a great help in understanding the SCIM concepts!

I'm looking at using your SCIM implementation for a very generic SCIM gateway, where the actual domain-specific work happens in separate services that listen to Service Bus events.

In this scenario, it would be valuable to
* on update events, also receive some kind of diff (maybe just a list of attribute paths that have changed)
* on remove events, get the `ExternalId`, since that will often be present in data structures on the event consumer side and can be used as a key. (Otherwise we have to store the "internal" SCIM Guid somewhere for mapping it to the ExternalId)

Would this fit in your design? If so, I can take a look at creating a PR for it.

**Comment by simpleidserver** [comment]
Hello,

You're right. SCIM solution can still be improved but we must be careful and we should not implement regressions :)

In the current implementation, the entire representation of a User is sent to a message broker.
**Advantage** : new subscribers are receiving the entire representation.
**Disadvantage** : More traffic, more difficult to update only a subset of data.

With your proposed solution, only modified data are sent to the message broker.
**Advantage** : subscribers know the changes and can properly update their data.
**Disadvantage** : new subscribers don't know all the properties but only the modified ones.

**Acceptable solution** : An EventStore can be used to tackle both problems : persist all the domain events in an EventStore, new subscribers can read all the messages messages from the beginning.
What do you think ?

**Reply by JWMB** [reply]
The EventStore is an interesting suggestion, but I imagine it to be a bit of a niche scenario to need to replay all events rather than getting the current state?
My suggestion is to send the current state, as is currently done, but also include a diff from the last state (probably in the JSON Patch format - you're already using that in other parts of the system, right?)

**Reply by simpleidserver** [reply]
Indeed in our scenario the EventStore usage can be overkill.
JSON PATCH request can be passed with the current state. Therefore your proposed solution can be implemented in SID.
You're free to make a pull request with those changes :) All contributions are welcomed
